src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -22,48 +26,83 @@ protected NamingStyleDiagnosticAnalyzerBase()
 
         // Applicable SymbolKind list is limited due to https://github.com/dotnet/roslyn/issues/8753. 
         // We would prefer to respond to the names of all symbols.
-        private static readonly ImmutableArray<SymbolKind> _symbolKinds = new[]
-            {
-                SymbolKind.Event,
-                SymbolKind.Field,
-                SymbolKind.Method,
-                SymbolKind.NamedType,
-                SymbolKind.Namespace,
-                SymbolKind.Property
-            }.ToImmutableArray();
+        private static readonly ImmutableArray<SymbolKind> _symbolKinds = ImmutableArray.Create(
+            SymbolKind.Event,
+            SymbolKind.Field,
+            SymbolKind.Method,
+            SymbolKind.NamedType,
+            SymbolKind.Namespace,
+            SymbolKind.Property);
 
         public bool OpenFileOnly(Workspace workspace) => true;
 
         protected override void InitializeWorker(AnalysisContext context)
-            => context.RegisterSymbolAction(SymbolAction, _symbolKinds);
+            => context.RegisterCompilationStartAction(CompilationStartAction);
 
-        private void SymbolAction(SymbolAnalysisContext context)
+        private void CompilationStartAction(CompilationStartAnalysisContext context)
         {
-            var namingStyleRules = context.GetNamingStyleRulesAsync().GetAwaiter().GetResult();
-            if (namingStyleRules == null)
+            var idToCachedResult = new ConcurrentDictionary<Guid, ConcurrentDictionary<string, string>>(
+                concurrencyLevel: 2, capacity: 0);
+            var preferencesToRules = new ConcurrentDictionary<NamingStylePreferences, NamingStyleRules>(
```

Comment:
>preferencesToRules [](start = 16, length = 18)

Couldn't we remove this and instead cache globally in calls to context.GetNamingStylePreferencesAsync()? I also don't see why preferencesToRules needs to be in a CompilationStartAction.  PerDocument options have nothing to do with compilations today